propagation.go

```diff
@@ -0,0 +1,103 @@
+package opentracing
+
+///////////////////////////////////////////////////////////////////////////////
+// CORE PROPAGATION INTERFACES:
+///////////////////////////////////////////////////////////////////////////////
+
+// PropagationInjector is responsible for injecting Span instances in a manner
+// suitable for propagation via a format-specific "carrier" object. Typically
+// the injection will take place across an RPC boundary, but message queues and
+// other IPC mechanisms are also reasonable places to use a
+// PropagationInjector.
+//
+// See PropagationExtractor and Span.PropagationInjectorForFormat.
+type PropagationInjector interface {
+	InjectSpan(span Span, carrier interface{})
+}
+
+// PropagationExtractor is responsible for extracting Span instances from an
+// format-specific "carrier" object. Typically the extraction will take place
+// on the server side of an RPC boundary, but message queues and other IPC
+// mechanisms are also reasonable places to use a PropagationExtractor.
+//
+// See PropagationInjector and Tracer.PropagationExtractorForFormat.
+type PropagationExtractor interface {
+	ExtractSpan(operationName string, carrier interface{}) (Span, error)
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// BUILTIN PROPAGATION FORMATS:
+///////////////////////////////////////////////////////////////////////////////
+
+// BuiltinPropagationFormat is the shared type for builtin OpenTracing in-band
+// propagation formats.
+type BuiltinPropagationFormat int
+
+const (
+	// PROPAGATION_FORMAT_SPLIT_BINARY encodes the Span in a BinaryCarrier
+	// instance.
+	//
+	// The `carrier` for injection and extraction must be a `*BinaryCarrier`
+	// instance.
+	PROPAGATION_FORMAT_SPLIT_BINARY BuiltinPropagationFormat = iota
+
+	// PROPAGATION_FORMAT_SPLIT_BINARY encodes the Span in a TextCarrier
+	// instance.
+	//
+	// The `carrier` for injection and extraction must be a `*TextCarrier`
+	// instance.
+	PROPAGATION_FORMAT_SPLIT_TEXT
+
+	// PROPAGATION_FORMAT_GO_HTTP_HEADER encodes the Span into a Go http.Header
+	// instance (both the tracer state and any Trace Attributes).
+	//
+	// The `carrier` for both injection and extraction must be an http.Header
+	// instance.
+	PROPAGATION_FORMAT_GO_HTTP_HEADER
```

Comment:
I don't like the idea of OpenTracing adding tons of builtin formats, but maybe we could stick with standard-library representations of super-common things like HTTP headers, etc. Or, say, maybe `opentracing-ruby` would have a builtin format for some Rails type just because so much of Ruby code is actually Rails. Et cetera... It would be an ongoing compromise, and IPC codebase maintainers would be strongly encouraged to fall back on one of the general-purpose text/binary formats as a contingency.